File: scripts/bitrix-merge-crm-sources.py
"""
Bitrix24 CRM "Источник" cleanup: merges duplicate directory entries
(Telegram/MAX/WhatsApp variants) into three canonical values, across
leads, deals, contacts and companies.

Setup:
    pip install requests
    export BITRIX_WEBHOOK_URL="https://your-portal.bitrix24.ru/rest/1/xxxxxxxxxxxxxxxx/"

Usage (always run in this order):
    python bitrix-merge-crm-sources.py --plan
        Prints the current directory + how many records reference each
        value to merge. Makes no changes.

    python bitrix-merge-crm-sources.py --apply
        Reassigns SOURCE_ID on all matching leads/deals/contacts/companies
        to the canonical value, and renames the three canonical directory
        entries. Does NOT delete the now-empty old directory entries.

    python bitrix-merge-crm-sources.py --cleanup
        Deletes old directory entries that no longer have any records
        pointing at them (re-checks before each delete; refuses to delete
        anything still referenced).

Webhook needs scopes: crm (read+write) covering leads, deals, contacts,
companies and the CRM status directory.
"""

import argparse
import os
import sys
import time
from urllib.parse import urlencode

import requests

WEBHOOK = os.environ.get("BITRIX_WEBHOOK_URL", "").rstrip("/") + "/"

ENTITY_TYPE_IDS = {
    "lead": 1,
    "deal": 2,
    "contact": 3,
    "company": 4,
}

# canonical name -> {rename_from: old NAME to become the canonical one,
#                     merge_from: other NAMEs whose records move to the canonical one}
MERGE_PLAN = {
    "Telegram": {
        "rename_from": "Telegram-бот",
        "merge_from": [
            "Инвайт ТГ",
            "WAZZUP: Telegram - Открытая линия",
            "Tgapi Клюев Андрей ТГ",
            "Tgapi 79310096002",
            "TG Bot kluevzabotabot",
            "Tgapi ТГ 3332",
            "Telegram kluvand_bot",
        ],
    },
    "MAX": {
        "rename_from": "MAX-бот",
        "merge_from": [
            "Инвайт МАКС",
            "MAX - МАКС",
            "WAZZUP: Max - Открытая линия 2",
            "Max 79307073332",
            "Max МАКС 79307073332",
            "Maxbot id525603925717_bot",
        ],
    },
    "WhatsApp": {
        "rename_from": "Whatsapp 79307073332",
        "merge_from": [],
    },
}


def call(method, params=None):
    resp = requests.post(WEBHOOK + method, json=params or {}, timeout=30)
    resp.raise_for_status()
    data = resp.json()
    if "error" in data:
        raise RuntimeError(f"{method} failed: {data.get('error_description', data['error'])}")
    return data


def batch(cmd):
    """cmd: dict[key] -> (method, params dict). Runs in chunks of 50, halt=0."""
    items = list(cmd.items())
    results = {}
    errors = {}
    for i in range(0, len(items), 50):
        chunk = dict(items[i:i + 50])
        encoded = {k: f"{m}?{urlencode(p, doseq=True)}" for k, (m, p) in chunk.items()}
        data = call("batch", {"halt": 0, "cmd": encoded})
        results.update(data["result"]["result"])
        errors.update(data["result"].get("result_error", {}))
        time.sleep(0.3)
    return results, errors


def get_source_directory():
    """Returns dict[NAME] -> list of {ID, STATUS_ID, NAME} (names can repeat)."""
    items = call("crm.status.list", {
        "filter": {"ENTITY_ID": "SOURCE"},
        "order": {"SORT": "ASC"},
    })["result"]
    by_name = {}
    for item in items:
        by_name.setdefault(item["NAME"], []).append(item)
    return by_name, items


def find_items_by_source(entity_type_id, source_codes, select=("id",)):
    found = []
    start = 0
    while True:
        data = call("crm.item.list", {
            "entityTypeId": entity_type_id,
            "select": list(select),
            "filter": {"@sourceId": source_codes},
            "start": start,
        })
        items = data["result"].get("items", [])
        found.extend(items)
        nxt = data.get("next")
        if nxt is None:
            break
        start = nxt
    return found


def plan():
    by_name, _ = get_source_directory()
    print("=== Текущий справочник Источник ===")
    for name, entries in by_name.items():
        for e in entries:
            print(f"  ID={e['ID']:<5} STATUS_ID={e['STATUS_ID']:<20} NAME={name}")
    print()

    for canonical, cfg in MERGE_PLAN.items():
        rename_from = cfg["rename_from"]
        merge_from = cfg["merge_from"]
        print(f"--- {canonical} (переименовать из «{rename_from}») ---")
        if rename_from not in by_name:
            print(f"  !! НЕ НАЙДЕНО в справочнике: {rename_from}")
            continue
        canonical_codes = [e["STATUS_ID"] for e in by_name[rename_from]]
        for old_name in merge_from:
            entries = by_name.get(old_name)
            if not entries:
                print(f"  !! НЕ НАЙДЕНО в справочнике: {old_name}")
                continue
            old_codes = [e["STATUS_ID"] for e in entries]
            for entity_name, entity_type_id in ENTITY_TYPE_IDS.items():
                try:
                    items = find_items_by_source(entity_type_id, old_codes)
                    if items:
                        print(f"  {old_name}: {len(items)} записей в {entity_name}")
                except RuntimeError as e:
                    print(f"  {old_name}: {entity_name} недоступен для sourceId ({e})")
        print()


def apply_merge():
    by_name, _ = get_source_directory()

    for canonical, cfg in MERGE_PLAN.items():
        rename_from = cfg["rename_from"]
        merge_from = cfg["merge_from"]

        if rename_from not in by_name:
            print(f"!! Пропускаю {canonical}: не найден исходник «{rename_from}»")
            continue

        canonical_entry = by_name[rename_from][0]
        canonical_id = canonical_entry["ID"]
        canonical_code = canonical_entry["STATUS_ID"]

        old_codes = []
        for old_name in merge_from:
            for e in by_name.get(old_name, []):
                old_codes.append(e["STATUS_ID"])

        if old_codes:
            print(f"=== {canonical}: переношу записи с {old_codes} на {canonical_code} ===")
            for entity_name, entity_type_id in ENTITY_TYPE_IDS.items():
                try:
                    items = find_items_by_source(entity_type_id, old_codes)
                except RuntimeError as e:
                    print(f"  {entity_name}: пропущено ({e})")
                    continue
                if not items:
                    continue
                print(f"  {entity_name}: обновляю {len(items)} записей")
                cmd = {
                    f"u{item['id']}": (
                        "crm.item.update",
                        {
                            "entityTypeId": entity_type_id,
                            "id": item["id"],
                            "fields": {"sourceId": canonical_code},
                        },
                    )
                    for item in items
                }
                _, errors = batch(cmd)
                if errors:
                    print(f"    ошибки: {errors}")

        print(f"Переименовываю directory ID={canonical_id} «{rename_from}» -> «{canonical}»")
        call("crm.status.update", {"id": canonical_id, "fields": {"NAME": canonical}})
        print()

    print("Готово. Старые пункты справочника пока не удалены — запустите --cleanup после проверки.")


def cleanup():
    by_name, _ = get_source_directory()
    for canonical, cfg in MERGE_PLAN.items():
        for old_name in cfg["merge_from"]:
            for e in list(by_name.get(old_name, [])):
                still_referenced = False
                for entity_name, entity_type_id in ENTITY_TYPE_IDS.items():
                    try:
                        items = find_items_by_source(entity_type_id, [e["STATUS_ID"]])
                    except RuntimeError:
                        continue
                    if items:
                        still_referenced = True
                        print(f"!! {old_name} (ID={e['ID']}) всё ещё используется в {entity_name} ({len(items)}), не удаляю")
                if not still_referenced:
                    print(f"Удаляю пустой пункт справочника: {old_name} (ID={e['ID']})")
                    call("crm.status.delete", {"id": e["ID"]})


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--plan", action="store_true")
    group.add_argument("--apply", action="store_true")
    group.add_argument("--cleanup", action="store_true")
    args = parser.parse_args()

    if not os.environ.get("BITRIX_WEBHOOK_URL"):
        sys.exit("Задайте переменную окружения BITRIX_WEBHOOK_URL")

    if args.plan:
        plan()
    elif args.apply:
        apply_merge()
    elif args.cleanup:
        cleanup()

File: scripts/bitrix-merge-crm-sources.ts
// Bitrix24 CRM "Источник" cleanup: merges duplicate directory entries
// (Telegram/MAX/WhatsApp variants) into three canonical values, across
// leads, deals, contacts and companies.
//
// Setup:
//   export BITRIX_WEBHOOK_URL="https://your-portal.bitrix24.ru/rest/1/xxxxxxxxxxxxxxxx/"
//
// Usage (always run in this order):
//   bun run scripts/bitrix-merge-crm-sources.ts --plan
//     Prints the current directory + how many records reference each
//     value to merge. Makes no changes.
//
//   bun run scripts/bitrix-merge-crm-sources.ts --apply
//     Reassigns SOURCE_ID on all matching leads/deals/contacts/companies
//     to the canonical value, and renames the three canonical directory
//     entries. Does NOT delete the now-empty old directory entries.
//     Prints a full plan and requires typing "ДА" before writing anything.
//
//   bun run scripts/bitrix-merge-crm-sources.ts --cleanup
//     Deletes old directory entries that no longer have any records
//     pointing at them (re-checks before each delete; refuses to delete
//     anything still referenced). Also requires typing "ДА".
//
// Webhook needs scopes: crm (read+write) covering leads, deals, contacts,
// companies and the CRM status directory.

import { createInterface } from "node:readline/promises";

const WEBHOOK = `${(process.env.BITRIX_WEBHOOK_URL || "").replace(/\/+$/, "")}/`;

const ENTITY_TYPE_IDS: Record<string, number> = {
	lead: 1,
	deal: 2,
	contact: 3,
	company: 4,
};

type MergePlanEntry = { renameFrom: string; mergeFrom: string[] };

const MERGE_PLAN: Record<string, MergePlanEntry> = {
	Telegram: {
		renameFrom: "Telegram-бот",
		mergeFrom: [
			"Инвайт ТГ",
			"WAZZUP: Telegram - Открытая линия",
			"Tgapi Клюев Андрей ТГ",
			"Tgapi 79310096002",
			"TG Bot kluevzabotabot",
			"Tgapi ТГ 3332",
			"Telegram kluvand_bot",
		],
	},
	MAX: {
		renameFrom: "MAX-бот",
		mergeFrom: [
			"Инвайт МАКС",
			"MAX - МАКС",
			"WAZZUP: Max - Открытая линия 2",
			"Max 79307073332",
			"Max МАКС 79307073332",
			"Maxbot id525603925717_bot",
		],
	},
	WhatsApp: {
		renameFrom: "Whatsapp 79307073332",
		mergeFrom: [],
	},
};

type StatusEntry = { ID: string | number; STATUS_ID: string; NAME: string };
type BitrixItem = { id: string | number; [key: string]: unknown };
type BitrixResponse = {
	result?: unknown;
	error?: string;
	error_description?: string;
	next?: number;
};

async function call(
	method: string,
	params: Record<string, unknown> = {},
): Promise<BitrixResponse> {
	const resp = await fetch(WEBHOOK + method, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(params),
	});
	const data = (await resp.json()) as BitrixResponse;
	if (data.error) {
		throw new Error(
			`${method} failed: ${data.error_description || data.error}`,
		);
	}
	return data;
}

// Bitrix batch cmd values are query strings, not JSON, so nested objects
// (e.g. fields: {sourceId: ...}) must become fields[sourceId]=... pairs -
// URLSearchParams alone won't flatten nested objects for you.
function flattenParams(
	params: Record<string, unknown>,
	prefix = "",
): [string, string][] {
	const pairs: [string, string][] = [];
	for (const [k, v] of Object.entries(params)) {
		const key = prefix ? `${prefix}[${k}]` : k;
		if (v !== null && typeof v === "object" && !Array.isArray(v)) {
			pairs.push(...flattenParams(v as Record<string, unknown>, key));
		} else if (Array.isArray(v)) {
			v.forEach((item, i) => {
				if (item !== null && typeof item === "object") {
					pairs.push(...flattenParams(item, `${key}[${i}]`));
				} else {
					pairs.push([`${key}[${i}]`, String(item)]);
				}
			});
		} else {
			pairs.push([key, String(v)]);
		}
	}
	return pairs;
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

async function batchCall(
	cmd: Record<string, [string, Record<string, unknown>]>,
) {
	const entries = Object.entries(cmd);
	const results: Record<string, unknown> = {};
	const errors: Record<string, unknown> = {};
	for (let i = 0; i < entries.length; i += 50) {
		const chunk = entries.slice(i, i + 50);
		const encoded: Record<string, string> = {};
		for (const [key, [method, params]] of chunk) {
			const qs = new URLSearchParams(flattenParams(params)).toString();
			encoded[key] = `${method}?${qs}`;
		}
		const data = await call("batch", { halt: 0, cmd: encoded });
		const batchResult = data.result as {
			result?: Record<string, unknown>;
			result_error?: Record<string, unknown>;
		};
		Object.assign(results, batchResult.result);
		Object.assign(errors, batchResult.result_error || {});
		await sleep(300);
	}
	return { results, errors };
}

async function getSourceDirectory(): Promise<Map<string, StatusEntry[]>> {
	const data = await call("crm.status.list", {
		filter: { ENTITY_ID: "SOURCE" },
		order: { SORT: "ASC" },
	});
	const byName = new Map<string, StatusEntry[]>();
	for (const item of data.result as StatusEntry[]) {
		if (!byName.has(item.NAME)) byName.set(item.NAME, []);
		byName.get(item.NAME)?.push(item);
	}
	return byName;
}

// entityTypeId -> whether it has a "sourceId" field, filled in lazily.
// Never assume - contacts/companies may not have this field on every
// portal, and an unsupported "@sourceId" filter could in theory be
// ignored by the API instead of erroring, which would turn a "merge these
// 12 records" update into "update every contact in the CRM".
const sourceFieldSupport = new Map<number, boolean>();

async function entitySupportsSource(entityTypeId: number): Promise<boolean> {
	if (!sourceFieldSupport.has(entityTypeId)) {
		const data = await call("crm.item.fields", { entityTypeId });
		const fields =
			(data.result as { fields?: Record<string, unknown> } | undefined)
				?.fields || {};
		sourceFieldSupport.set(entityTypeId, "sourceId" in fields);
	}
	const supported = sourceFieldSupport.get(entityTypeId);
	if (supported === undefined)
		throw new Error(
			`entitySupportsSource: missing cache entry for ${entityTypeId}`,
		);
	return supported;
}

// Returns null (not []) if this entity type has no sourceId field at all,
// so callers can tell "nothing to merge" apart from "can't check".
async function findItemsBySource(
	entityTypeId: number,
	sourceCodes: string[],
	select: string[] = ["id"],
): Promise<BitrixItem[] | null> {
	if (!(await entitySupportsSource(entityTypeId))) return null;

	const found: BitrixItem[] = [];
	let start = 0;
	const seenStarts = new Set<number>();
	while (true) {
		const data = await call("crm.item.list", {
			entityTypeId,
			select,
			filter: { "@sourceId": sourceCodes },
			start,
		});
		const result =
			(data.result as { items?: BitrixItem[]; next?: number } | undefined) ||
			{};
		const items = Array.isArray(result.items) ? result.items : [];
		found.push(...items);
		const nxt = data.next ?? result.next ?? null;
		if (nxt === null || nxt === undefined || seenStarts.has(nxt)) break;
		seenStarts.add(nxt);
		start = nxt;
	}
	return found;
}

async function confirm(promptText: string): Promise<boolean> {
	const rl = createInterface({ input: process.stdin, output: process.stdout });
	const answer = (await rl.question(promptText)).trim();
	rl.close();
	return answer === "ДА";
}

function printTargetPortal() {
	try {
		console.log(`Портал: ${new URL(WEBHOOK).host}\n`);
	} catch {
		console.log("!! BITRIX_WEBHOOK_URL не похож на валидный URL\n");
	}
}

function warnDuplicateCanonicalNames(byName: Map<string, StatusEntry[]>) {
	for (const cfg of Object.values(MERGE_PLAN)) {
		const entries = byName.get(cfg.renameFrom);
		if (entries && entries.length > 1) {
			console.log(
				`!! В справочнике ${entries.length} записей с именем «${cfg.renameFrom}» — использую только ID=${entries[0].ID}, остальные (${entries
					.slice(1)
					.map((e) => e.ID)
					.join(", ")}) не будут учтены`,
			);
		}
	}
}

async function plan() {
	printTargetPortal();
	const byName = await getSourceDirectory();
	warnDuplicateCanonicalNames(byName);
	console.log("=== Текущий справочник Источник ===");
	for (const [name, entries] of byName) {
		for (const e of entries) {
			console.log(
				`  ID=${String(e.ID).padEnd(5)} STATUS_ID=${String(e.STATUS_ID).padEnd(20)} NAME=${name}`,
			);
		}
	}
	console.log();

	console.log("=== Поддержка поля sourceId по типам сущностей ===");
	for (const [entityName, entityTypeId] of Object.entries(ENTITY_TYPE_IDS)) {
		const supported = await entitySupportsSource(entityTypeId);
		console.log(
			`  ${entityName}: ${supported ? "да" : "НЕТ — будет пропущен"}`,
		);
	}
	console.log();

	for (const [canonical, cfg] of Object.entries(MERGE_PLAN)) {
		console.log(`--- ${canonical} (переименовать из «${cfg.renameFrom}») ---`);
		if (!byName.has(cfg.renameFrom)) {
			console.log(`  !! НЕ НАЙДЕНО в справочнике: ${cfg.renameFrom}`);
			continue;
		}
		for (const oldName of cfg.mergeFrom) {
			const entries = byName.get(oldName);
			if (!entries) {
				console.log(`  !! НЕ НАЙДЕНО в справочнике: ${oldName}`);
				continue;
			}
			const oldCodes = entries.map((e) => e.STATUS_ID);
			for (const [entityName, entityTypeId] of Object.entries(
				ENTITY_TYPE_IDS,
			)) {
				const items = await findItemsBySource(entityTypeId, oldCodes);
				if (items?.length) {
					console.log(`  ${oldName}: ${items.length} записей в ${entityName}`);
				}
			}
		}
		console.log();
	}
}

type WorkItem = {
	canonical: string;
	canonicalId: string | number;
	canonicalCode: string;
	entityName: string;
	entityTypeId: number;
	items: BitrixItem[];
};

async function collectApplyPlan(byName: Map<string, StatusEntry[]>) {
	const work: WorkItem[] = [];
	const skipped: string[] = [];

	for (const [canonical, cfg] of Object.entries(MERGE_PLAN)) {
		if (!byName.has(cfg.renameFrom)) {
			skipped.push(`${canonical}: не найден исходник «${cfg.renameFrom}»`);
			continue;
		}
		const canonicalEntry = byName.get(cfg.renameFrom)?.[0];
		if (!canonicalEntry) {
			skipped.push(`${canonical}: не найден исходник «${cfg.renameFrom}»`);
			continue;
		}
		const canonicalId = canonicalEntry.ID;
		const canonicalCode = canonicalEntry.STATUS_ID;

		const oldCodes: string[] = [];
		for (const oldName of cfg.mergeFrom) {
			const entries = byName.get(oldName);
			if (!entries) {
				skipped.push(
					`${canonical}: не найден дубль «${oldName}» (пропускаю только его)`,
				);
				continue;
			}
			oldCodes.push(...entries.map((e) => e.STATUS_ID));
		}

		for (const [entityName, entityTypeId] of Object.entries(ENTITY_TYPE_IDS)) {
			if (!(await entitySupportsSource(entityTypeId))) continue;
			const items = oldCodes.length
				? await findItemsBySource(entityTypeId, oldCodes)
				: [];
			work.push({
				canonical,
				canonicalId,
				canonicalCode,
				entityName,
				entityTypeId,
				items: items || [],
			});
		}
	}
	return { work, skipped };
}

const MAX_SANE_UPDATE_COUNT = 2000;

async function applyMerge() {
	printTargetPortal();
	const byName = await getSourceDirectory();
	warnDuplicateCanonicalNames(byName);
	const { work, skipped } = await collectApplyPlan(byName);

	console.log("=== План изменений ===");
	for (const msg of skipped) console.log(`  !! ${msg}`);
	let total = 0;
	for (const w of work) {
		if (w.items.length) {
			console.log(
				`  ${w.canonical}: ${w.items.length} записей в ${w.entityName} -> sourceId=${w.canonicalCode}`,
			);
			total += w.items.length;
		}
	}
	for (const [canonical, cfg] of Object.entries(MERGE_PLAN)) {
		console.log(`  Переименовать «${cfg.renameFrom}» -> «${canonical}»`);
	}
	console.log(`\nВсего записей будет обновлено: ${total}`);

	if (total > MAX_SANE_UPDATE_COUNT) {
		console.log(
			`\n!! Это подозрительно много (> ${MAX_SANE_UPDATE_COUNT}) для ожидаемого объёма дублей мессенджеров.`,
			"Похоже на ошибку в справочнике или фильтре, а не на реальный список дублей. Прерываю без изменений — проверьте --plan вручную.",
		);
		return;
	}

	if (!(await confirm("\nВведите ДА для применения изменений: "))) {
		console.log("Отменено, ничего не изменено.");
		return;
	}

	for (const w of work) {
		if (!w.items.length) continue;
		console.log(
			`${w.canonical} / ${w.entityName}: обновляю ${w.items.length} записей`,
		);
		const cmd: Record<string, [string, Record<string, unknown>]> = {};
		for (const item of w.items) {
			cmd[`u${item.id}`] = [
				"crm.item.update",
				{
					entityTypeId: w.entityTypeId,
					id: item.id,
					fields: { sourceId: w.canonicalCode },
				},
			];
		}
		const { errors } = await batchCall(cmd);
		if (Object.keys(errors).length) console.log("  ошибки:", errors);
	}

	for (const [canonical, cfg] of Object.entries(MERGE_PLAN)) {
		if (!byName.has(cfg.renameFrom)) continue;
		const canonicalId = byName.get(cfg.renameFrom)?.[0]?.ID;
		if (canonicalId === undefined) continue;
		console.log(
			`Переименовываю directory ID=${canonicalId} «${cfg.renameFrom}» -> «${canonical}»`,
		);
		try {
			await call("crm.status.update", {
				id: canonicalId,
				fields: { NAME: canonical },
			});
		} catch (e) {
			console.log(
				`  !! не удалось переименовать: ${e instanceof Error ? e.message : e}`,
			);
		}
	}

	console.log(
		"\nГотово. Старые пункты справочника пока не удалены — запустите --cleanup после проверки.",
	);
}

async function cleanup() {
	printTargetPortal();
	const byName = await getSourceDirectory();
	warnDuplicateCanonicalNames(byName);
	const toDelete: { oldName: string; id: string | number }[] = [];

	for (const [, cfg] of Object.entries(MERGE_PLAN)) {
		for (const oldName of cfg.mergeFrom) {
			for (const e of byName.get(oldName) || []) {
				let stillReferenced = false;
				for (const [entityName, entityTypeId] of Object.entries(
					ENTITY_TYPE_IDS,
				)) {
					if (!(await entitySupportsSource(entityTypeId))) continue;
					const items = await findItemsBySource(entityTypeId, [e.STATUS_ID]);
					if (items?.length) {
						stillReferenced = true;
						console.log(
							`!! ${oldName} (ID=${e.ID}) всё ещё используется в ${entityName} (${items.length}), не удаляю`,
						);
					}
				}
				if (!stillReferenced) toDelete.push({ oldName, id: e.ID });
			}
		}
	}

	if (!toDelete.length) {
		console.log("Нечего удалять.");
		return;
	}

	console.log("\nБудут удалены пустые пункты справочника:");
	for (const { oldName, id } of toDelete)
		console.log(`  ${oldName} (ID=${id})`);

	if (!(await confirm("\nВведите ДА для удаления: "))) {
		console.log("Отменено, ничего не удалено.");
		return;
	}

	for (const { oldName, id } of toDelete) {
		console.log(`Удаляю: ${oldName} (ID=${id})`);
		try {
			await call("crm.status.delete", { id });
		} catch (e) {
			console.log(
				`  !! не удалось удалить: ${e instanceof Error ? e.message : e}`,
			);
		}
	}
}

async function main() {
	if (!process.env.BITRIX_WEBHOOK_URL) {
		console.error("Задайте переменную окружения BITRIX_WEBHOOK_URL");
		process.exit(1);
	}

	const args = process.argv.slice(2);
	if (args.includes("--plan")) await plan();
	else if (args.includes("--apply")) await applyMerge();
	else if (args.includes("--cleanup")) await cleanup();
	else {
		console.error("Укажите один из флагов: --plan, --apply, --cleanup");
		process.exit(1);
	}
}

main().catch((e) => {
	console.error(e instanceof Error ? e.message : e);
	process.exit(1);
});
